core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java

```diff
@@ -0,0 +1,130 @@
+ * to destination database
+ */
+public class FileCopyRecordWriter extends RecordWriter<NullWritable, AbstractFileMetadata> {
+  private FileSystem destFileSystem;
+  private String basePath;
+  private Boolean enableOverwrite;
+  private int bufferSize;
+  // buffer size defaults to 1 MB
+  private static final int defaultBufferSize = 2 << 20;
+  private static final Logger LOG = LoggerFactory.getLogger(FileCopyRecordWriter.class);
+
+  public FileCopyRecordWriter(Configuration conf)
+    throws Exception {
+    this.destFileSystem = FileSystem.get(conf);
+    this.basePath = conf.get(FileCopyOutputFormat.BASE_PATH);
+    this.enableOverwrite = conf.getBoolean(FileCopyOutputFormat.ENABLE_OVERWRITE, false);
+    this.bufferSize = conf.getInt(FileCopyOutputFormat.BUFFER_SIZE, defaultBufferSize);
+    this.sourceFilesystemMap = new HashMap<>();
+  }
+
+  // source filesystems
+  private Map<AbstractFileMetadata.Credentials, FileSystem> sourceFilesystemMap;
+
+  @Override
+  public void write(NullWritable key, AbstractFileMetadata fileMetadata) throws IOException, InterruptedException {
+    // get source database connection
+    AbstractFileMetadata.Credentials sourceCredentials = fileMetadata.getCredentials();
+    if (!sourceFilesystemMap.containsKey(sourceCredentials)) {
+      sourceFilesystemMap.put(sourceCredentials, getSourceFilesystemFromCredentials(sourceCredentials));
+    }
+    FileSystem sourceFilesystem = sourceFilesystemMap.get(sourceCredentials);
+
+    // construct file paths for source and destination
+    Path srcPath = new Path(fileMetadata.fullPath);
+    Path destPath = new Path(basePath + "/" + fileMetadata.basePath);
+    FsPermission permission = new FsPermission(fileMetadata.getPermission());
+
+    if (fileMetadata.isFolder) {
+      // create an empty folder and return
+      if (!destFileSystem.exists(destPath) && sourceFilesystem.isDirectory(srcPath)) {
+        destFileSystem.mkdirs(destPath, permission);
+      }
+      return;
+    } else if (!sourceFilesystem.exists(srcPath)) {
+      // file doesn't exist in source, return immediately
+      return;
+    }
+
+    // immediately return if we dont want to overwrite and file exists in destination
+    if (!enableOverwrite && destFileSystem.exists(destPath)) {
+      return;
+    }
+
+    // data streaming
+    FSDataInputStream inputStream = sourceFilesystem.open(srcPath);
+    FSDataOutputStream outputStream = FileSystem.create(destFileSystem, destPath, permission);
+    byte[] buf = new byte[bufferSize];
+    int len;
+    while ((len = inputStream.read(buf)) >= 0) {
+      outputStream.write(buf, 0, len);
+    }
+    inputStream.close();
+    outputStream.close();
+  }
+
+  @Override
+  public void close(TaskAttemptContext taskAttemptContext) throws IOException, InterruptedException {
+    /*
+     * TODO: investigate when to close the destination filesystem and the source filesystems
```

Comment:
I see. I guess this means we have some questions to figure out:
1. If we turn caching off, is it okay to have 100 splits all try to initiate connection to the filesystem at the same time?
2. If we leave caching on, does that mean file transmission resources (like bandwidth and other things) would be divided among the splits?

Maybe we can address these problems in a future PR.